lib/cloud_controller/upload_buildpack.rb

```diff
@@ -47,8 +50,31 @@ def upload_buildpack(buildpack, bits_file_path, new_filename)
       true
     end
 
+    def extract_stack_from_buildpack(bits_file_path)
+      bits_file_path = bits_file_path.path if bits_file_path.respond_to?(:path)
+      output, _, status = Open3.capture3('unzip', '-p', bits_file_path, 'manifest.yml')
+      YAML.safe_load(output).dig('stack') if status.success?
+    end
+
     private
 
+    def determine_new_stack(buildpack, bits_file_path)
+      extracted_stack = extract_stack_from_buildpack(bits_file_path)
+      if extracted_stack.to_s != '' && !Stack.where(name: extracted_stack).first
+        raise CloudController::Errors::ApiError.new_from_details('BuildpackStackDoesNotExist', extracted_stack)
+      end
+      new_stack = [extracted_stack, buildpack.stack, Stack.default.name].find { |s| s.to_s != '' && s.to_s != 'unknown' }
```

Comment:
The `s.to_s != '' && s.to_s != 'unknown'` can be moved into the Stack model and given a meaningful name.

Reply:
Agree with your comments prior and here regarding `unknown`. Will look at solving this, if not with `Stack.defined_name?` then something similar.

We'll go with `present?` since it's possible someone could inject an empty string as well.